src/Core/Tokens/Token.cs

```diff
@@ -26,11 +27,28 @@ public Token RemovePrefix(string expectedPrefix)
         return new Token(_token[expectedPrefix.Length..]);
     }
 
-    public Token ProtectWith(IDataProtector dataProtector) =>
-        new(dataProtector.Protect(ToString()));
 
-    public Token UnprotectWith(IDataProtector dataProtector) =>
-        new(dataProtector.Unprotect(ToString()));
+    public Token ProtectWith<T>(IDataProtector dataProtector, ILogger<T> logger)
+    {
+        logger.LogDebug("Protecting token: {0}", this);
+        return new(dataProtector.Protect(ToString()));
+    }
+
+    public Token UnprotectWith<T>(IDataProtector dataProtector, ILogger<T> logger)
+    {
+        var unprotected = "";
+        try
+        {
+            unprotected = dataProtector.Unprotect(ToString());
+        }
+        catch (Exception e)
+        {
+            logger.LogInformation(e, "Failed to unprotect token: {0}", this);
+            throw;
+        }
+        logger.LogDebug("Unprotected token: {0} to {1}", this, unprotected);
+        return new(dataProtector.Unprotect(ToString()));
```

Comment:
Should we use the `unprotected` variable here instead of doing this again?

Reply:
yup, thanks